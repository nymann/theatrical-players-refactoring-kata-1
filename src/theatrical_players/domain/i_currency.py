from abc import ABC, abstractmethod

class ICurrency(ABC):
    @abstractmethod
    def format(self, amount: float) -> str:
        pass
