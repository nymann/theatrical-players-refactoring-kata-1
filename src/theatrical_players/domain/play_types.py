from abc import ABC, abstractmethod

class IPlayType(ABC):

    @abstractmethod
    def volume_credits(self, audience) -> int:
        pass

    @abstractmethod
    def bonus(self, audience: int) -> int:
        pass

class PlayType(IPlayType):
    def bonus(self, audience: int) -> int:
        raise NotImplementedError(audience)

    def volume_credits(self, audience: int) -> int:
        return max(audience - 30, 0)

    @classmethod
    def from_str(cls, name_type: str):
        try:
            return PLAY_TYPES[name_type]
        except KeyError as e:
            raise ValueError(f"unknown type: {name_type}") from e

class Tragedy(PlayType):
    def bonus(self, audience: int) -> int:
        amount = 40000
        min = 30
        if audience > min:
            amount += 1000 * (audience - min)
        return amount


class Comedy(PlayType):
    def bonus(self, audience: int) -> int:
        amount = 30000
        min = 20
        if audience > min:
            amount += 10000 + 500 * (audience - min)
        amount += 300 * audience
        return amount

    def volume_credits(self, audience: int) -> int:
        v = super().volume_credits(audience)
        v += audience // 5
        return v

PLAY_TYPES = {"comedy": Comedy(), "tragedy": Tragedy()}
