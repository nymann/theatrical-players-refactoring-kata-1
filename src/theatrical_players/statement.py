import math
import pydantic
from typing import Dict, List, Any

class Play(pydantic.BaseModel):
    play_id: str
    name: str
    play_type: str

    @classmethod
    def from_json(cls, play_id: str, play: Dict[str, str]):
        return cls(play_id=play_id, name=play["name"], play_type=play["type"])


class Performance(pydantic.BaseModel):
    play: Play
    audience: int

class Invoice(pydantic.BaseModel):
    customer: str
    performances: List[Performance]

class Statement(object):
    def __init__(self, invoice: Invoice) -> None:
        self.invoice = invoice

    @classmethod
    def from_json(cls, invoice: Dict[str, Any], plays: Dict[str, Dict[str, str]]):
        temp_plays: Dict[str, Play] = {}
        for play_id, play in plays.items():
            temp_plays[play_id] = Play.from_json(play_id=play_id, play=play)

        performances: List[Performance] = []
        for performance in invoice["performances"]:
            play_id = performance["playID"]
            play = temp_plays[play_id]
            audience = performance["audience"]
            performances.append(Performance(play=play, audience=audience))

        transformed_invoice = Invoice(customer=invoice["customer"], performances=performances)
        
        return cls(transformed_invoice)

    def calculate(self):
        total_amount: float = 0
        volume_credits: float = 0
        result: str = f'Statement for {self.invoice.customer}\n'
        for perf in self.invoice.performances:
            play = perf.play
            if play.play_type == "tragedy":
                this_amount = 40000
                if perf.audience > 30:
                    this_amount += 1000 * (perf.audience - 30)
            elif play.play_type == "comedy":
                this_amount = 30000
                if perf.audience > 20:
                    this_amount += 10000 + 500 * (perf.audience - 20)
    
                this_amount += 300 * perf.audience
    
            else:
                raise ValueError(f'unknown type: {play.play_type}')
    
            # add volume credits
            volume_credits += max(perf.audience - 30, 0)
            # add extra credit for every ten comedy attendees
            if "comedy" == play.play_type:
                volume_credits += math.floor(perf.audience / 5)
            # print line for this order
            result += f' {play.name}: {self._format_as_dollars(this_amount/100)} ({perf.audience} seats)\n'
            total_amount += this_amount
    
        result += f"Amount owed is {self._format_as_dollars(total_amount/100)}\n"
        result += f"You earned {volume_credits} credits\n"
        return result

    def _format_as_dollars(self, amount: float) -> str:
        return f"${amount:0,.2f}"
