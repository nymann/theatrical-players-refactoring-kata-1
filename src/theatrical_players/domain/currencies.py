from theatrical_players.domain.i_currency import ICurrency


class Usd(ICurrency):
    def format(self, amount: float) -> str:
        return f"${amount/100:0,.2f}"
