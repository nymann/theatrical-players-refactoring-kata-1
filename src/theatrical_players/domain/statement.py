from typing import Dict, List, Any
from theatrical_players.data.order import Order
from theatrical_players.data.invoice import Invoice
from theatrical_players.data.play import Play
from theatrical_players.data.performance import Performance
from theatrical_players.domain.play_types import PlayType

class Statement(object):
    def __init__(self, invoice: Invoice) -> None:
        self.invoice: Invoice = invoice
        self.volume_credits: float = 0
        self.total_amount_cents: float = 0
        self.orders: List[Order] = []
        for performance in self.invoice.performances:
            play = performance.play
            play_type = play.play_type
            this_amount_cents = play_type.bonus(audience=performance.audience)
            self.volume_credits += play_type.volume_credits(audience=performance.audience)
            order = Order(name=play.name, audience=performance.audience, amount=this_amount_cents)
            self.orders.append(order)
            self.total_amount_cents += this_amount_cents


    @classmethod
    def from_json(cls, invoice: Dict[str, Any], plays: Dict[str, Dict[str, str]]):
        temp_plays: Dict[str, Play] = {}
        for play_id, play in plays.items():
            play_type = PlayType.from_str(play["type"])
            temp_plays[play_id] = Play(play_id=play_id, name=play["name"], play_type=play_type)

        performances: List[Performance] = []
        for performance in invoice["performances"]:
            play_id = performance["playID"]
            play = temp_plays[play_id]
            audience = performance["audience"]
            performances.append(Performance(play=play, audience=audience))

        transformed_invoice = Invoice(customer=invoice["customer"], performances=performances)
        
        return cls(transformed_invoice)


    def _format_as_dollars(self, amount: float) -> str:
        return f"${amount:0,.2f}"
