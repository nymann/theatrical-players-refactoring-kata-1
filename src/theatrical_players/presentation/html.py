from theatrical_players.domain.i_currency import ICurrency
from theatrical_players.domain.currencies import Usd
from theatrical_players.presentation.presentation import Presentation

class Html(Presentation):
    def __str__(self, currency: ICurrency = Usd()):
        customer = self.statement.invoice.customer
        html = self._header(f"Statement | {customer}")
        html += self._style()
        body_text: str = f"<h1>Statement for {customer}</h1>\n"
        body_text += "<table>\n<tr>\n\t<th>Name</th>\n\t<th>Amount ($)</th>\n\t<th>Audience (seats)</th>\n</tr>\n"
        audience_sum: int = 0
        for order in self.statement.orders:
            order_amount = currency.format(order.amount)
            audience_sum += order.audience
            body_text += "<tr>\n"
            body_text += f"\t<td>{order.name}</td>\n"
            body_text += f"\t<td>{order_amount}</td>\n"
            body_text += f"\t<td>{order.audience}</td>\n"
            body_text += "</tr>\n"
        total_amount = currency.format(self.statement.total_amount_cents)
        body_text += f"<tr>\n\t<td><strong>Total</strong></td>\n\t<td><strong>{total_amount}</strong></td>\n\t<td><strong>{audience_sum}</strong></td>\n</tr>\n"
        body_text += "<table>\n"
        body_text += f"<h3>You earned {self.statement.volume_credits} credits</h3>"
        html += self._body(body_text)
        html += "</html>\n"
        return html

    def _header(self, title: str):
        return f"<!doctype html>\n<html>\n<head>\n\t<title>{title}</title>\n</head>\n"

    def _body(self, text: str):
        return f"<body>\n{text}\n</body>\n"

    def _style(self):
        return "<style>\ntable, th, td {\n\tborder: 1px solid black;\n}\n</style>\n"
