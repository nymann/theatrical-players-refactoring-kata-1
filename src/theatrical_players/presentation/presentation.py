from theatrical_players.presentation.i_presentation import IPresentation
from theatrical_players.domain.statement import Statement

class Presentation(IPresentation):
    def __init__(self, statement: Statement):
        self.statement = statement

    def __str__(self):
        raise NotImplementedError()
