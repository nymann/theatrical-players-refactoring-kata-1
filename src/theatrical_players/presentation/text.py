from theatrical_players.presentation.presentation import Presentation

class Text(Presentation):
    def __str__(self):
        result: str = f'Statement for {self.statement.invoice.customer}\n'
        for order in self.statement.orders:
            result += f' {order.name}: {self.statement._format_as_dollars(order.amount/100)} ({order.audience} seats)\n'
        result += f"Amount owed is {self.statement._format_as_dollars(self.statement.total_amount_cents/100)}\n"
        result += f"You earned {self.statement.volume_credits} credits\n"
        return result
