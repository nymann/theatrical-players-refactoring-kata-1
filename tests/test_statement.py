import json

from approvaltests.utils import get_adjacent_file
from approvaltests import verify
import pytest

from theatrical_players.domain.statement import Statement
from theatrical_players.presentation.text import Text

def test_example_statement():
    with open(get_adjacent_file("invoice.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("plays.json")) as f:
        plays = json.loads(f.read())
    statement = Statement.from_json(invoice=invoice, plays=plays)
    presentation = Text(statement)
    verify(str(presentation))


def test_statement_with_new_play_types():
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    statement = Statement.from_json(invoice=invoice, plays=plays)
    presentation = Text(statement)
    verify(str(presentation))
