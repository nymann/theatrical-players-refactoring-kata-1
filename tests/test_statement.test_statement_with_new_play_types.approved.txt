Statement for BigCoII
 Henry V: $1,940.00 (53 seats)
 As You Like It: $0.00 (55 seats)
Amount owed is $1,940.00
You earned 48 credits
